File: go/position_stats.go
package iriq

import "strconv"

// DefaultMaxValuesPerPosition mirrors PositionStats::DEFAULT_MAX_VALUES.
// Existing tracked values still receive increments after the cap is hit;
// only NEW distinct values are dropped.
const DefaultMaxValuesPerPosition = 5_000

// PositionStats holds rolling frequency counts for a single
// (host, prefix-shape, position). Value cardinality is capped so a
// high-entropy position (UUIDs, timestamps) doesn't grow without bound.
//
// Numeric* fields track min/max/sum over numeric-typed observations so
// Cluster.ParamType can promote integer positions whose values land in a
// recognizable range (e.g. years in 1900..2100) to a more specific type.
type PositionStats struct {
	ValueCounts map[string]int
	TypeCounts  map[SegmentType]int
	Total       int
	MaxValues   int

	NumericCount int
	NumericMin   float64
	NumericMax   float64
	NumericSum   float64
}

func NewPositionStats(maxValues int) *PositionStats {
	if maxValues <= 0 {
		maxValues = DefaultMaxValuesPerPosition
	}
	return &PositionStats{
		ValueCounts: map[string]int{},
		TypeCounts:  map[SegmentType]int{},
		MaxValues:   maxValues,
	}
}

func (p *PositionStats) Observe(value string, t SegmentType) {
	p.Total++
	p.TypeCounts[t]++
	if _, present := p.ValueCounts[value]; present || len(p.ValueCounts) < p.MaxValues {
		p.ValueCounts[value]++
	}
	p.recordNumeric(value, t)
}

func (p *PositionStats) recordNumeric(value string, t SegmentType) {
	if t != TypeInteger && t != TypeFloat {
		return
	}
	n, err := strconv.ParseFloat(value, 64)
	if err != nil {
		return
	}
	if p.NumericCount == 0 || n < p.NumericMin {
		p.NumericMin = n
	}
	if p.NumericCount == 0 || n > p.NumericMax {
		p.NumericMax = n
	}
	p.NumericCount++
	p.NumericSum += n
}

// NumericAvg returns the mean of numeric observations (0 when none).
func (p *PositionStats) NumericAvg() float64 {
	if p.NumericCount == 0 {
		return 0
	}
	return p.NumericSum / float64(p.NumericCount)
}

func (p *PositionStats) Cardinality() int { return len(p.ValueCounts) }

// VariableFraction returns the share of observations whose classifier-derived
// type was variable (anything other than :literal).
func (p *PositionStats) VariableFraction(c *SegmentClassifier) float64 {
	if p.Total == 0 {
		return 0
	}
	v := 0
	for t, n := range p.TypeCounts {
		if c.Variable(t) {
			v += n
		}
	}
	return float64(v) / float64(p.Total)
}

func (p *PositionStats) ValueFraction(value string) float64 {
	if p.Total == 0 {
		return 0
	}
	return float64(p.ValueCounts[value]) / float64(p.Total)
}

// DominantType returns the SegmentType with the largest count in TypeCounts.
// On ties, returns the lexicographically smallest type so the answer is
// deterministic across runs (Go's map iteration order is randomized;
// Ruby's max_by returns the first-inserted max, which is also deterministic
// per-process — and lexicographic tie-break keeps Ruby↔Go behavior aligned
// without requiring insertion order).
func (p *PositionStats) DominantType() SegmentType {
	var dom SegmentType
	var domN int
	first := true
	for t, n := range p.TypeCounts {
		if first || n > domN || (n == domN && string(t) < string(dom)) {
			dom = t
			domN = n
			first = false
		}
	}
	return dom
}

File: rust/iriq/src/position_stats.rs
use crate::classifier::{SegmentClassifier, SegmentType};
use std::collections::HashMap;

pub const DEFAULT_MAX_VALUES_PER_POSITION: usize = 5_000;

#[derive(Debug, Clone, PartialEq)]
#[non_exhaustive]
pub struct PositionStats {
    pub value_counts: HashMap<String, usize>,
    pub type_counts: HashMap<SegmentType, usize>,
    pub total: usize,
    pub max_values: usize,
    pub numeric_count: usize,
    pub numeric_min: f64,
    pub numeric_max: f64,
    pub numeric_sum: f64,
}

impl PositionStats {
    pub fn new(max_values: usize) -> Self {
        let cap = if max_values == 0 {
            DEFAULT_MAX_VALUES_PER_POSITION
        } else {
            max_values
        };
        PositionStats {
            value_counts: HashMap::new(),
            type_counts: HashMap::new(),
            total: 0,
            max_values: cap,
            numeric_count: 0,
            numeric_min: 0.0,
            numeric_max: 0.0,
            numeric_sum: 0.0,
        }
    }

    pub fn observe(&mut self, value: &str, t: SegmentType) {
        self.total += 1;
        *self.type_counts.entry(t).or_insert(0) += 1;
        let present = self.value_counts.contains_key(value);
        if present || self.value_counts.len() < self.max_values {
            *self.value_counts.entry(value.to_string()).or_insert(0) += 1;
        }
        self.record_numeric(value, t);
    }

    fn record_numeric(&mut self, value: &str, t: SegmentType) {
        if t != SegmentType::Integer && t != SegmentType::Float {
            return;
        }
        // A 310+ digit value overflows to ±inf; it still counts as an
        // observation, but would poison min/max/avg (and JSON) as a range stat.
        let Some(n) = value.parse::<f64>().ok().filter(|n| n.is_finite()) else {
            return;
        };
        if self.numeric_count == 0 || n < self.numeric_min {
            self.numeric_min = n;
        }
        if self.numeric_count == 0 || n > self.numeric_max {
            self.numeric_max = n;
        }
        self.numeric_count += 1;
        self.numeric_sum += n;
    }

    pub fn numeric_avg(&self) -> f64 {
        if self.numeric_count == 0 {
            0.0
        } else {
            self.numeric_sum / (self.numeric_count as f64)
        }
    }

    pub fn cardinality(&self) -> usize {
        self.value_counts.len()
    }

    pub fn variable_fraction(&self, c: &SegmentClassifier) -> f64 {
        if self.total == 0 {
            return 0.0;
        }
        let v: usize = self
            .type_counts
            .iter()
            .filter(|(t, _)| c.variable(**t))
            .map(|(_, n)| *n)
            .sum();
        (v as f64) / (self.total as f64)
    }

    pub fn value_fraction(&self, value: &str) -> f64 {
        if self.total == 0 {
            return 0.0;
        }
        let n = *self.value_counts.get(value).unwrap_or(&0);
        (n as f64) / (self.total as f64)
    }

    /// SegmentType with the largest count. Ties broken lexicographically
    /// (by the type's display string) for cross-runtime determinism.
    pub fn dominant_type(&self) -> SegmentType {
        let mut best: Option<(SegmentType, usize)> = None;
        for (&t, &n) in &self.type_counts {
            best = match best {
                None => Some((t, n)),
                Some((bt, bn)) => {
                    if n > bn || (n == bn && t.as_str() < bt.as_str()) {
                        Some((t, n))
                    } else {
                        Some((bt, bn))
                    }
                }
            };
        }
        best.map(|(t, _)| t).unwrap_or(SegmentType::Literal)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn non_finite_numbers_count_as_observations_but_not_as_range_stats() {
        // 400 digits overflows f64 to ±inf.
        let huge = "1".repeat(400);
        let mut stats = PositionStats::new(0);
        stats.observe(&huge, SegmentType::Integer);
        assert_eq!(
            stats.numeric_count, 0,
            "an all-infinite position has no range"
        );

        stats.observe("1", SegmentType::Integer);
        stats.observe(&format!("-{huge}.5"), SegmentType::Float);
        stats.observe("3.5", SegmentType::Float);
        assert_eq!(stats.total, 4);
        assert_eq!(stats.type_counts[&SegmentType::Integer], 2);
        assert_eq!(stats.cardinality(), 4);
        assert_eq!(
            (stats.numeric_count, stats.numeric_min, stats.numeric_max),
            (2, 1.0, 3.5)
        );
        assert_eq!(stats.numeric_avg(), 2.25);
    }
}
